src/util/match.h

```diff
@@ -21,4 +21,18 @@ template<class... Ts> struct match : Ts... { using Ts::operator()...; };
 // explicit deduction guide (not needed as of C++20)
 template<class... Ts> match(Ts...) -> match<Ts...>;
 
+// A wrapper around two-argument `std::visit` that reverses the arguments, putting the
+// value to be visited first. This is normally used as:
+//
+// examine(specimen, match {
+//     ...
+// })
+//
+// The return type is inferred as it would be for `std::visit`.
+#define examine(specimen, visitor...) (std::visit(single_arg(visitor), (specimen)))
```

Comment:
The design of the C preprocessor basically assumed that commas in an expression would always be surrounded by parens, which is mostly true for C. But in C++ code it is common to have commas between braces or angle brackets, which causes the preprocessor to treat that code as multiple arguments when passed to a macro. In particular, uses of `match` with more than one alternative will always have this problem, due to the commas between the alternatives. The errors look something like:
```c++
./zcash/Address.hpp:109:17: error: too many arguments provided to function-like macro invocation
                [&](const libzcash::SaplingPaymentAddress &zaddr) {
                ^
./util/match.h:33:9: note: macro 'examine' defined here
#define examine(specimen, visitor) (std::visit((visitor), (specimen)))
        ^
```

Variadic macros parse their arguments in a way that allows commas within braces or angle brackets, so this just works.

Reply:
What about avoiding the macro with
```c++
template<class Variant, class Visitor>
constexpr decltype(auto) examine(Variant&& var, Visitor&& vis) {
    return std::visit(vis, var);
}
```
which should work the same at call sites.

Or even
```c++
template<class Variant, class... Visitors>
constexpr decltype(auto) match_visit(Variant&& var, Visitors&&... vis) {
    return std::visit(match { vis... }, var);
}
```
which would eliminate the nested `match`, like
```c++
    examine(changeAddr,
        [&](const CKeyID& keyId) {
            tChangeAddr = keyId;
        },
        [&](const CScriptID& scriptId) {
            tChangeAddr = scriptId;
        },
        [&](const libzcash::SaplingPaymentAddress& changeDest) {
            saplingChangeAddr = std::make_pair(ovk, changeDest);
        },
        [&](const libzcash::OrchardRawAddress& changeDest) {
            orchardChangeAddr = std::make_pair(ovk, changeDest);
        });
```

Reply:
@daira I missed that you had a comment shortly before mine that mentioned a similar approach. You’re right – we can’t infer `R` (until C++20), but using `decltype(auto)` works and is how `visit` is actually implemented in Clang for C++17.

Having `R` in C++20 will be nice, because it should eliminate a lot of the `-> some_type` annotations on the individual closures.

Reply:
I couldn't get the second variant with `std::visit(match { vis... }, var)` to work, but the first one does. Thanks! Force-pushed to use that.

Reply:
Btw, I think that `examine` shouldn't be `constexpr` because it doesn't meet [the requirements for a `constexpr` function](https://en.cppreference.com/w/cpp/language/constexpr#constexpr_function). I don't know why clang wasn't complaining.